core/vm: faster* bitvec implementation

lmittmann opened the pull request:
This PR adds a new `bitvec` implementation that is faster in practice and most existing benchmarks.

The core implementation idea is still the same (mark PUSH data with a 1-bit), but by using `[]uint32` instead of `[]byte` the maximum number of slice elements that need to be touched is reduced to two.

For now both implementations exit for easy comparability and testing. Should you decide to merge I would do some cleanup first.

## Benchmark Results
```
$ benchstat -col /v bitvec.bench
                              │     old      │                 new                 │
                              │    sec/op    │   sec/op     vs base                │
JumpdestAnalysis_rand-8         1133.2µ ± 1%   542.6µ ± 1%  -52.11% (p=0.000 n=10)
JumpdestAnalysis_weth9-8        1000.2n ± 0%   890.0n ± 0%  -11.02% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH1-8       821.8µ ± 1%   780.0µ ± 1%   -5.08% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH2-8       588.2µ ± 2%   876.0µ ± 1%  +48.94% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH3-8       463.9µ ± 0%   658.2µ ± 0%  +41.87% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH4-8       374.8µ ± 1%   526.7µ ± 1%  +40.54% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH5-8       314.4µ ± 1%   439.7µ ± 0%  +39.86% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH6-8       288.1µ ± 2%   376.4µ ± 1%  +30.66% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH7-8       232.8µ ± 1%   331.2µ ± 0%  +42.30% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH8-8       265.7µ ± 0%   293.6µ ± 0%  +10.48% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH9-8       290.3µ ± 2%   264.9µ ± 1%   -8.74% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH10-8      394.0µ ± 0%   240.6µ ± 0%  -38.94% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH11-8      266.8µ ± 0%   221.2µ ± 0%  -17.08% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH12-8      274.6µ ± 3%   204.4µ ± 0%  -25.54% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH13-8      259.2µ ± 2%   189.8µ ± 0%  -26.75% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH14-8      296.1µ ± 2%   177.5µ ± 0%  -40.03% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH15-8      200.0µ ± 1%   166.4µ ± 0%  -16.79% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH16-8      155.3µ ± 0%   156.6µ ± 0%   +0.88% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH17-8      177.0µ ± 1%   148.0µ ± 0%  -16.35% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH18-8      227.9µ ± 1%   140.6µ ± 1%  -38.32% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH19-8      173.5µ ± 1%   133.7µ ± 1%  -22.96% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH20-8      166.1µ ± 3%   127.4µ ± 1%  -23.35% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH21-8      166.2µ ± 2%   121.8µ ± 1%  -26.71% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH22-8      188.0µ ± 2%   117.0µ ± 1%  -37.77% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH23-8      147.7µ ± 4%   111.8µ ± 1%  -24.30% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH24-8      139.7µ ± 2%   107.2µ ± 0%  -23.30% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH25-8      177.9µ ± 3%   103.5µ ± 1%  -41.84% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH26-8      171.1µ ± 4%   100.1µ ± 1%  -41.48% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH27-8     188.84µ ± 6%   96.44µ ± 2%  -48.93% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH28-8     163.66µ ± 4%   92.95µ ± 1%  -43.21% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH29-8     158.16µ ± 6%   89.74µ ± 0%  -43.26% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH30-8     184.84µ ± 3%   94.65µ ± 6%  -48.80% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH31-8     157.25µ ± 3%   84.39µ ± 2%  -46.33% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH32-8     122.27µ ± 2%   55.22µ ± 2%  -54.84% (p=0.000 n=10)
JumpdestOpAnalysis/JUMPDEST-8    562.5µ ± 0%   533.9µ ± 5%   -5.08% (p=0.007 n=10)
JumpdestOpAnalysis/STOP-8        569.8µ ± 2%   530.3µ ± 0%   -6.93% (p=0.000 n=10)
geomean                          217.3µ        171.9µ       -20.89%
```
<details>
<summary>Old Results</summary>
<pre>
$ benchstat -col /v bitvec.bench
                              │     old      │                 new                  │
                              │    sec/op    │    sec/op     vs base                │
JumpdestAnalysis_rand-8         1111.5µ ± 0%    508.5µ ± 1%  -54.25% (p=0.000 n=20)
JumpdestAnalysis_weth9-8        1006.0n ± 1%    808.0n ± 1%  -19.68% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH1-8       844.5µ ± 3%   1550.8µ ± 1%  +83.65% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH2-8       588.2µ ± 2%   1028.3µ ± 1%  +74.81% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH3-8       459.3µ ± 1%    773.2µ ± 3%  +68.32% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH4-8       425.8µ ± 2%    618.2µ ± 2%  +45.18% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH5-8       349.6µ ± 1%    490.2µ ± 1%  +40.23% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH6-8       322.1µ ± 1%    450.8µ ± 1%  +39.93% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH7-8       233.2µ ± 3%    393.2µ ± 1%  +68.60% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH8-8       268.2µ ± 5%    341.3µ ± 2%  +27.29% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH9-8       309.1µ ± 1%    311.7µ ± 0%   +0.85% (p=0.009 n=20)
JumpdestOpAnalysis/PUSH10-8      410.0µ ± 4%    279.5µ ± 2%  -31.82% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH11-8      278.7µ ± 3%    265.1µ ± 1%   -4.89% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH12-8      309.4µ ± 1%    225.2µ ± 1%  -27.21% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH13-8      287.4µ ± 1%    226.8µ ± 1%  -21.08% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH14-8      292.1µ ± 2%    213.8µ ± 1%  -26.81% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH15-8      203.7µ ± 5%    196.1µ ± 3%   -3.73% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH16-8      168.8µ ± 1%    186.8µ ± 1%  +10.66% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH17-8      190.1µ ± 1%    168.9µ ± 5%  -11.17% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH18-8      251.5µ ± 1%    156.5µ ± 2%  -37.78% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH19-8      185.0µ ± 3%    159.9µ ± 1%  -13.57% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH20-8      170.4µ ± 1%    152.4µ ± 1%  -10.60% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH21-8      173.1µ ± 6%    146.2µ ± 1%  -15.57% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH22-8      212.5µ ± 1%    137.0µ ± 4%  -35.52% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH23-8      156.2µ ± 1%    123.1µ ± 1%  -21.17% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH24-8      151.3µ ± 1%    129.5µ ± 2%  -14.38% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH25-8      181.8µ ± 3%    125.5µ ± 1%  -30.99% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH26-8      176.9µ ± 3%    120.9µ ± 1%  -31.69% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH27-8      199.0µ ± 1%    116.0µ ± 1%  -41.72% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH28-8      188.5µ ± 1%    108.6µ ± 2%  -42.43% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH29-8      182.2µ ± 1%    102.8µ ± 1%  -43.56% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH30-8      184.1µ ± 1%    105.6µ ± 1%  -42.64% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH31-8      172.3µ ± 3%    102.9µ ± 1%  -40.31% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH32-8     124.65µ ± 1%    99.15µ ± 1%  -20.46% (p=0.000 n=20)
JumpdestOpAnalysis/JUMPDEST-8    552.5µ ± 0%    401.3µ ± 3%  -27.37% (p=0.000 n=20)
JumpdestOpAnalysis/STOP-8        580.9µ ± 1%    409.6µ ± 1%  -29.49% (p=0.000 n=20)
geomean                          228.9µ         199.4µ       -12.89%
</pre>
</details>

holiman commented:
Hm, so the mean is faster, 

```
geomean                          228.9µ         199.4µ       -12.89%
``` 
But I don't think that's as important as keeping the worst-case in check. 
```
JumpdestOpAnalysis/PUSH1-8       844.5µ ± 3%   1550.8µ ± 1%  +83.65% (p=0.000 n=20)
```
To be safe against attacks, we need to ensure that the worst-case doesn't hurt us too much.

lmittmann commented:
> But I don't think that's as important as keeping the worst-case in check.

I see your point. Although if you look at this the other way around you have the same worst-case somewhere else:
```
                            │     new     │                 old                 │
                            │   sec/op    │   sec/op     vs base                │
JumpdestOpAnalysis/PUSH27-8   116.0µ ± 1%   199.0µ ± 1%  +71.58% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH28-8   108.6µ ± 2%   188.5µ ± 1%  +73.69% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH29-8   102.8µ ± 1%   182.2µ ± 1%  +77.17% (p=0.000 n=20)
JumpdestOpAnalysis/PUSH30-8   105.6µ ± 1%   184.1µ ± 1%  +74.34% (p=0.000 n=20)
```

FWIW I see an overall VM performance improvement of 2% in my real-mainnet-blockinterval benchmark. This includes additional overhead of my `w3vm` package, so the pure `core/vm` performance improvement should be even slightly better (https://github.com/lmittmann/w3/blob/main/w3vm/bench_test.go).

holiman commented:
A sequence of `PUSH1` is worst case:  `844.5µ < 116.0µ`. I don't understand how you can interpret it any other way ?

- A non-PUSH will shortcut the loop early (but also always do `n` spins in the loop for `n` ops)
- A HIGH push, like `PUSH32` will advance in steps of `32`, thus doing _way_ less spins in the loop
- A low push, like `PUSH1` will enter the heavy part of the loop every time, and although it will skip ahead, it will only skip ahead one step, so will do `n/2` spins in the loop for `n` ops).

lmittmann commented:
Yes, I see your point, in absolute numbers `PUSH1` is clearly the worst case. Maybe I manage to fix the worst case. The overall gain of 2% is quite nice IMO.

holiman commented:
Sure! I suspect that the map lookups are taking a toll.

lmittmann commented:
> Sure! I suspect that the map lookups are taking a toll.
No, that's an array lookup and actually an optimization.

lmittmann commented:
@holiman I updated the implementation and the benchmark results. The mean performance improved by another 8%. `PUSH1` is now slightly faster than the original implementation. The new worst-case is `PUSH2`. However, its performance is close to the performance of `PUSH1` in the old implementation.

lmittmann commented:
@holiman did you had a chance to look at the updated results?

holiman commented:
Not yet, no. I'm curious about 
```
JumpdestAnalysis_rand-8         1133.2µ ± 1%   542.6µ ± 1%  -52.11% (p=0.000 n=10)
```
I don't see any reason for a 52% speedup in the general/random case. Sometimes there are just weird side-effects depending on certain options that the compiler happens to do, or not do, which don't really correspond to actual code features.

lmittmann commented:
I don't really understand the degree of performance increase for the random case either. However, the overall performance improvement comes from the reduced number of slice accesses and thus the fewer number of slice bound checks inserted by the compiler. See https://go.godbolt.org/z/hrE5e7naP

lmittmann commented:
hi @holiman, any chance of getting this performance improvement merged?

namiloh commented:
Well, sure at some point, but it is still in draft-mode, for starters

lmittmann commented:
if the code is okay so far from your end I will finalize the PR and drop the duplicated code I kept for testing and benchmarking

namiloh commented:
I'd drop it, yes, but: put all benchmarking changes in one commit, and code-changes in a second. That makes it easy to compare by just jumping forward/back

lmittmann commented:
@holiman I dropped the old implementation for the new. The old benchmark logic was not changed to keep them comparable (I added a few new though). The test cases where translated from the old impl to the new one.

* old code with additional benchmarks: [04593bb](https://github.com/ethereum/go-ethereum/pull/28984/commits/04593bb71a6219a8560dd8e593709962339900b8)
* new code: [071bb70](https://github.com/ethereum/go-ethereum/pull/28984/commits/071bb7082732ab59e4684e3617953d4f6a470748)

holiman commented:
Here are my numbers. 
The `1200k` is empty code, just `STOP` opcodes. It becomes 30% slower, whereas the totally random data becomes almost twice as fast. The weth-code drops 30% performance. 

```
name                      old time/op    new time/op    delta
JumpdestAnalysis_1200k-8    1.19ms ±14%    1.71ms ± 7%  +44.30%  (p=0.000 n=10+9)
JumpdestAnalysis_rand-8     2.90ms ±13%    1.53ms ±15%  -47.45%  (p=0.000 n=10+10)
JumpdestAnalysis_weth9-8    2.34µs ±57%    3.10µs ±31%  +32.24%  (p=0.029 n=10+10)

name                      old speed      new speed      delta
JumpdestAnalysis_1200k-8  1.04GB/s ±16%  0.72GB/s ± 8%  -31.05%  (p=0.000 n=10+9)
JumpdestAnalysis_rand-8    425MB/s ±12%   814MB/s ±14%  +91.50%  (p=0.000 n=10+10)
JumpdestAnalysis_weth9-8  1.43GB/s ±41%  1.04GB/s ±41%  -27.28%  (p=0.029 n=10+10)
```
As for the detailed opcodes: 
```
user@work go-ethereum]$ benchstat analysis.ops.1  analysis.ops.2
name                           old time/op    new time/op    delta
JumpdestOpAnalysis/PUSH1-8       2.43ms ±44%    2.16ms ±30%      ~     (p=0.529 n=10+10)
JumpdestOpAnalysis/PUSH2-8       1.95ms ±26%    2.67ms ±29%   +36.85%  (p=0.009 n=10+10)
JumpdestOpAnalysis/PUSH3-8       1.66ms ±44%    1.91ms ±27%      ~     (p=0.436 n=10+10)
JumpdestOpAnalysis/PUSH4-8       1.25ms ±28%    1.58ms ±24%      ~     (p=0.063 n=10+10)
JumpdestOpAnalysis/PUSH5-8        974µs ±31%    1363µs ±32%   +39.89%  (p=0.007 n=10+10)
JumpdestOpAnalysis/PUSH6-8        905µs ±22%    1090µs ±26%      ~     (p=0.052 n=10+10)
JumpdestOpAnalysis/PUSH7-8        828µs ±24%     915µs ±32%      ~     (p=0.631 n=10+10)
JumpdestOpAnalysis/PUSH8-8        948µs ±27%     721µs ±12%   -23.98%  (p=0.010 n=10+9)
JumpdestOpAnalysis/PUSH9-8        935µs ±27%     759µs ±27%   -18.84%  (p=0.035 n=10+10)
JumpdestOpAnalysis/PUSH10-8      1.14ms ± 0%    0.74ms ±30%   -34.91%  (p=0.000 n=8+10)
JumpdestOpAnalysis/PUSH11-8       885µs ±36%     610µs ±32%   -31.11%  (p=0.005 n=10+10)
JumpdestOpAnalysis/PUSH12-8       846µs ±27%     547µs ±33%   -35.28%  (p=0.000 n=10+10)
JumpdestOpAnalysis/PUSH13-8       794µs ±33%     586µs ±29%   -26.12%  (p=0.011 n=10+10)
JumpdestOpAnalysis/PUSH14-8       883µs ± 1%     479µs ±36%   -45.75%  (p=0.000 n=9+10)
JumpdestOpAnalysis/PUSH15-8       811µs ± 2%     527µs ±26%   -35.02%  (p=0.000 n=8+10)
JumpdestOpAnalysis/PUSH16-8       520µs ±29%     484µs ±26%      ~     (p=0.089 n=10+10)
JumpdestOpAnalysis/PUSH17-8       520µs ±33%     455µs ±29%      ~     (p=0.218 n=10+10)
JumpdestOpAnalysis/PUSH18-8       580µs ±24%     420µs ±31%   -27.67%  (p=0.035 n=10+10)
JumpdestOpAnalysis/PUSH19-8       461µs ± 7%     489µs ± 0%    +6.01%  (p=0.010 n=8+8)
JumpdestOpAnalysis/PUSH20-8       567µs ±28%     411µs ±27%   -27.53%  (p=0.015 n=10+10)
JumpdestOpAnalysis/PUSH21-8       562µs ±26%     351µs ±27%   -37.53%  (p=0.000 n=10+10)
JumpdestOpAnalysis/PUSH22-8       610µs ± 1%     383µs ±31%   -37.23%  (p=0.000 n=8+10)
JumpdestOpAnalysis/PUSH23-8       480µs ±22%     333µs ±27%   -30.57%  (p=0.004 n=10+10)
JumpdestOpAnalysis/PUSH24-8       538µs ±31%     342µs ±27%   -36.43%  (p=0.000 n=10+10)
JumpdestOpAnalysis/PUSH25-8       545µs ±32%     280µs ±36%   -48.62%  (p=0.000 n=10+10)
JumpdestOpAnalysis/PUSH26-8       615µs ±23%     283µs ±30%   -54.04%  (p=0.000 n=10+10)
JumpdestOpAnalysis/PUSH27-8       426µs ±12%     278µs ±28%   -34.87%  (p=0.000 n=10+10)
JumpdestOpAnalysis/PUSH28-8       506µs ±27%     268µs ±28%   -46.98%  (p=0.000 n=10+10)
JumpdestOpAnalysis/PUSH29-8       503µs ±32%     277µs ±30%   -44.96%  (p=0.000 n=10+10)
JumpdestOpAnalysis/PUSH30-8       429µs ±40%     271µs ±30%   -36.93%  (p=0.000 n=10+10)
JumpdestOpAnalysis/PUSH31-8       447µs ±30%     310µs ± 1%   -30.57%  (p=0.000 n=10+8)
JumpdestOpAnalysis/PUSH32-8       407µs ±39%     191µs ±37%   -53.20%  (p=0.000 n=10+10)
JumpdestOpAnalysis/JUMPDEST-8    1.34ms ±53%    1.55ms ±42%      ~     (p=0.105 n=10+10)
JumpdestOpAnalysis/STOP-8        1.08ms ±52%    1.43ms ±40%   +31.85%  (p=0.029 n=10+10)

name                           old speed      new speed      delta
JumpdestOpAnalysis/PUSH1-8      545MB/s ±65%   602MB/s ±28%      ~     (p=0.529 n=10+10)
JumpdestOpAnalysis/PUSH2-8      659MB/s ±30%   481MB/s ±34%   -26.97%  (p=0.009 n=10+10)
JumpdestOpAnalysis/PUSH3-8      839MB/s ±39%   664MB/s ±32%      ~     (p=0.436 n=10+10)
JumpdestOpAnalysis/PUSH4-8     1.04GB/s ±32%  0.80GB/s ±27%      ~     (p=0.063 n=10+10)
JumpdestOpAnalysis/PUSH5-8     1.32GB/s ±28%  0.95GB/s ±40%   -28.09%  (p=0.007 n=10+10)
JumpdestOpAnalysis/PUSH6-8     1.39GB/s ±20%  1.17GB/s ±25%      ~     (p=0.052 n=10+10)
JumpdestOpAnalysis/PUSH7-8     1.53GB/s ±29%  1.40GB/s ±28%      ~     (p=0.631 n=10+10)
JumpdestOpAnalysis/PUSH8-8     1.35GB/s ±32%  1.66GB/s ±31%   +23.05%  (p=0.019 n=10+10)
JumpdestOpAnalysis/PUSH9-8     1.37GB/s ±32%  1.70GB/s ±26%   +24.40%  (p=0.035 n=10+10)
JumpdestOpAnalysis/PUSH10-8    1.08GB/s ± 0%  1.75GB/s ±35%   +62.48%  (p=0.000 n=8+10)
JumpdestOpAnalysis/PUSH11-8    1.46GB/s ±30%  2.08GB/s ±27%   +42.38%  (p=0.005 n=10+10)
JumpdestOpAnalysis/PUSH12-8    1.50GB/s ±32%  2.29GB/s ±26%   +52.68%  (p=0.000 n=10+10)
JumpdestOpAnalysis/PUSH13-8    1.62GB/s ±42%  2.18GB/s ±35%   +34.11%  (p=0.011 n=10+10)
JumpdestOpAnalysis/PUSH14-8    1.39GB/s ± 1%  2.70GB/s ±30%   +93.92%  (p=0.000 n=9+10)
JumpdestOpAnalysis/PUSH15-8    1.51GB/s ± 2%  2.41GB/s ±31%   +59.46%  (p=0.000 n=8+10)
JumpdestOpAnalysis/PUSH16-8    2.49GB/s ±35%  2.65GB/s ±29%      ~     (p=0.089 n=10+10)
JumpdestOpAnalysis/PUSH17-8    2.54GB/s ±40%  2.82GB/s ±36%      ~     (p=0.218 n=10+10)
JumpdestOpAnalysis/PUSH18-8    2.20GB/s ±26%  3.07GB/s ±38%   +39.20%  (p=0.035 n=10+10)
JumpdestOpAnalysis/PUSH19-8    2.67GB/s ± 7%  2.51GB/s ± 0%    -5.88%  (p=0.010 n=8+8)
JumpdestOpAnalysis/PUSH20-8    2.25GB/s ±34%  3.09GB/s ±32%   +37.20%  (p=0.015 n=10+10)
JumpdestOpAnalysis/PUSH21-8    2.26GB/s ±32%  3.65GB/s ±31%   +61.63%  (p=0.000 n=10+10)
JumpdestOpAnalysis/PUSH22-8    2.01GB/s ± 1%  3.33GB/s ±39%   +65.54%  (p=0.000 n=8+10)
JumpdestOpAnalysis/PUSH23-8    2.64GB/s ±22%  3.88GB/s ±30%   +46.75%  (p=0.004 n=10+10)
JumpdestOpAnalysis/PUSH24-8    2.35GB/s ±41%  3.72GB/s ±32%   +58.29%  (p=0.000 n=10+10)
JumpdestOpAnalysis/PUSH25-8    2.37GB/s ±40%  4.55GB/s ±29%   +92.15%  (p=0.000 n=10+10)
JumpdestOpAnalysis/PUSH26-8    2.04GB/s ±27%  4.55GB/s ±29%  +123.64%  (p=0.000 n=10+10)
JumpdestOpAnalysis/PUSH27-8    2.90GB/s ±12%  4.65GB/s ±28%   +60.77%  (p=0.000 n=10+10)
JumpdestOpAnalysis/PUSH28-8    2.51GB/s ±25%  4.81GB/s ±29%   +92.06%  (p=0.000 n=10+10)
JumpdestOpAnalysis/PUSH29-8    2.59GB/s ±38%  4.65GB/s ±36%   +79.68%  (p=0.000 n=10+10)
JumpdestOpAnalysis/PUSH30-8    3.04GB/s ±33%  4.73GB/s ±37%   +55.78%  (p=0.000 n=10+10)
JumpdestOpAnalysis/PUSH31-8    2.92GB/s ±35%  3.96GB/s ± 1%   +35.77%  (p=0.000 n=10+8)
JumpdestOpAnalysis/PUSH32-8    3.22GB/s ±54%  6.85GB/s ±49%  +112.96%  (p=0.000 n=10+10)
JumpdestOpAnalysis/JUMPDEST-8  1.06GB/s ±84%  0.89GB/s ±52%      ~     (p=0.105 n=10+10)
JumpdestOpAnalysis/STOP-8      1.25GB/s ±41%  0.94GB/s ±42%   -24.93%  (p=0.029 n=10+10)
```

The new worst-case is `PUSH2`, which is worse than the previous worstcase `PUSH1`. 

This PR has the drawback that it makes the memory consumption larger (esp for small codes), since it allocates `32 * len(code)/32+2)` instead of `len(code)/8+1+4)` bytes. So `64` bytes for a 1-byte contract, instead of `5`. 

This PR has the upside that it's a bit simpler code. But OTOH it's new code, whereas the existing is old and trusted. 

All in all, I'm not sure if it's worth it. Feels a bit like whack-a-mole with speeds, improve it a bit here and it becomes slower there. I am not convinced that this PR is overall an improvement.

lmittmann commented:
Interesting, you have completely different results than I do. On which CPU architecture are you?

I just reran the benchmarks at the two commits, and essentially PUSH1..8, and PUSH16 perform worse, everything else is better.

```
goos: linux
goarch: amd64
pkg: github.com/ethereum/go-ethereum/core/vm
cpu: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
                              │   old.bench   │              new.bench              │
                              │    sec/op     │   sec/op     vs base                │
JumpdestAnalysis_1200k-8         568.1µ ±  0%   537.5µ ± 0%   -5.38% (p=0.000 n=10)
JumpdestAnalysis_rand-8         1029.6µ ±  0%   519.6µ ± 3%  -49.53% (p=0.000 n=10)
JumpdestAnalysis_weth9-8        1036.5n ±  1%   959.9n ± 4%   -7.39% (p=0.000 n=10)
JumpdestHashing_1200k-8          2.740m ±  3%   2.709m ± 5%        ~ (p=0.247 n=10)
JumpdestOpAnalysis/PUSH1-8       796.0µ ±  1%   838.9µ ± 3%   +5.38% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH2-8       607.8µ ±  3%   980.6µ ± 1%  +61.35% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH3-8       453.3µ ±  3%   734.6µ ± 3%  +62.05% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH4-8       443.8µ ± 10%   622.6µ ± 4%  +40.27% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH5-8       354.2µ ±  2%   529.5µ ± 2%  +49.51% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH6-8       302.0µ ±  1%   447.3µ ± 2%  +48.12% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH7-8       249.0µ ±  1%   397.1µ ± 5%  +59.50% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH8-8       256.6µ ±  1%   328.7µ ± 2%  +28.09% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH9-8       309.1µ ±  1%   301.9µ ± 2%   -2.32% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH10-8      404.1µ ±  1%   272.2µ ± 1%  -32.64% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH11-8      284.8µ ±  1%   248.3µ ± 1%  -12.84% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH12-8      289.3µ ±  1%   233.9µ ± 4%  -19.17% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH13-8      261.7µ ±  1%   229.7µ ± 2%  -12.23% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH14-8      247.2µ ±  2%   217.2µ ± 2%  -12.14% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH15-8      213.9µ ±  1%   200.2µ ± 2%   -6.41% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH16-8      151.8µ ±  1%   191.5µ ± 1%  +26.14% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH17-8      184.9µ ±  2%   181.5µ ± 2%   -1.88% (p=0.019 n=10)
JumpdestOpAnalysis/PUSH18-8      246.0µ ±  1%   165.0µ ± 2%  -32.94% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH19-8      177.5µ ±  3%   151.6µ ± 3%  -14.56% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH20-8      188.6µ ±  4%   142.6µ ± 2%  -24.42% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH21-8      184.0µ ±  2%   136.8µ ± 1%  -25.63% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH22-8      211.5µ ±  1%   134.4µ ± 7%  -36.48% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH23-8      156.7µ ±  1%   139.4µ ± 1%  -11.07% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH24-8      152.4µ ±  1%   133.9µ ± 2%  -12.11% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH25-8      190.5µ ±  1%   128.1µ ± 1%  -32.75% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH26-8      185.3µ ±  1%   123.0µ ± 1%  -33.60% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH27-8      184.0µ ±  1%   120.0µ ± 2%  -34.79% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH28-8      181.8µ ±  2%   114.0µ ± 1%  -37.31% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH29-8      172.5µ ±  1%   105.4µ ± 3%  -38.89% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH30-8      168.6µ ±  2%   100.2µ ± 2%  -40.59% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH31-8      158.9µ ± 10%   102.3µ ± 3%  -35.61% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH32-8     129.09µ ±  1%   64.81µ ± 2%  -49.79% (p=0.000 n=10)
JumpdestOpAnalysis/JUMPDEST-8    580.7µ ±  2%   545.4µ ± 2%   -6.08% (p=0.000 n=10)
JumpdestOpAnalysis/STOP-8        581.0µ ±  1%   551.8µ ± 2%   -5.03% (p=0.000 n=10)
geomean                          246.7µ         218.4µ       -11.49%

                              │   old.bench   │               new.bench               │
                              │      B/s      │      B/s       vs base                │
JumpdestAnalysis_1200k-8        2.014Gi ±  0%    2.129Gi ± 0%   +5.69% (p=0.000 n=10)
JumpdestAnalysis_rand-8         1.112Gi ±  0%    2.202Gi ± 3%  +98.13% (p=0.000 n=10)
JumpdestAnalysis_weth9-8        2.806Gi ±  1%    3.031Gi ± 4%   +8.02% (p=0.000 n=10)
JumpdestHashing_1200k-8         427.9Mi ±  3%    432.6Mi ± 5%        ~ (p=0.247 n=10)
JumpdestOpAnalysis/PUSH1-8      1.438Gi ±  1%    1.364Gi ± 3%   -5.10% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH2-8      1.883Gi ±  3%    1.167Gi ± 1%  -38.02% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH3-8      2.525Gi ±  3%    1.558Gi ± 2%  -38.29% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH4-8      2.578Gi ± 11%    1.839Gi ± 4%  -28.68% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH5-8      3.231Gi ±  2%    2.161Gi ± 2%  -33.11% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH6-8      3.789Gi ±  1%    2.558Gi ± 2%  -32.49% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH7-8      4.596Gi ±  1%    2.882Gi ± 5%  -37.31% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH8-8      4.460Gi ±  1%    3.482Gi ± 2%  -21.93% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH9-8      3.703Gi ±  1%    3.791Gi ± 2%   +2.37% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH10-8     2.832Gi ±  1%    4.204Gi ± 1%  +48.45% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH11-8     4.018Gi ±  1%    4.610Gi ± 1%  +14.73% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH12-8     3.956Gi ±  1%    4.894Gi ± 4%  +23.71% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH13-8     4.372Gi ±  1%    4.981Gi ± 2%  +13.93% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH14-8     4.629Gi ±  2%    5.268Gi ± 2%  +13.82% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH15-8     5.349Gi ±  1%    5.716Gi ± 2%   +6.86% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH16-8     7.538Gi ±  1%    5.976Gi ± 1%  -20.72% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH17-8     6.188Gi ±  2%    6.306Gi ± 2%   +1.91% (p=0.019 n=10)
JumpdestOpAnalysis/PUSH18-8     4.651Gi ±  1%    6.936Gi ± 2%  +49.11% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH19-8     6.448Gi ±  3%    7.547Gi ± 3%  +17.03% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH20-8     6.067Gi ±  4%    8.026Gi ± 2%  +32.30% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH21-8     6.219Gi ±  2%    8.363Gi ± 1%  +34.47% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH22-8     5.410Gi ±  1%    8.518Gi ± 7%  +57.44% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH23-8     7.302Gi ±  1%    8.211Gi ± 1%  +12.45% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH24-8     7.509Gi ±  1%    8.544Gi ± 2%  +13.78% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH25-8     6.007Gi ±  1%    8.933Gi ± 1%  +48.71% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH26-8     6.177Gi ±  1%    9.302Gi ± 1%  +50.61% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH27-8     6.218Gi ±  1%    9.535Gi ± 2%  +53.35% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH28-8     6.295Gi ±  2%   10.042Gi ± 1%  +59.51% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH29-8     6.635Gi ±  2%   10.859Gi ± 3%  +63.65% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH30-8     6.788Gi ±  2%   11.426Gi ± 2%  +68.32% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH31-8     7.204Gi ±  9%   11.188Gi ± 3%  +55.30% (p=0.000 n=10)
JumpdestOpAnalysis/PUSH32-8     8.866Gi ±  1%   17.658Gi ± 2%  +99.18% (p=0.000 n=10)
JumpdestOpAnalysis/JUMPDEST-8   1.971Gi ±  2%    2.098Gi ± 2%   +6.47% (p=0.000 n=10)
JumpdestOpAnalysis/STOP-8       1.970Gi ±  1%    2.074Gi ± 2%   +5.30% (p=0.000 n=10)
geomean                         3.963Gi          4.478Gi       +12.98%
```

holiman commented:
> All in all, I'm not sure if it's worth it. Feels a bit like whack-a-mole with speeds, improve it a bit here and it becomes slower there. I am not convinced that this PR is overall an improvement.

I'm going to close this PR with this reasoning. Thanks for the work you put in!
